run-make-check.sh: ccache goodness for everyone

smithfarm opened the pull request:
Since run-make-check.sh already ensures that ccache is installed, it makes sense to let everyone benefit from the ccache tweaks introduced by 4cb5a590537a9caaf61db42ce8ea123d2ab961f3.

For now I'm retaining the Jenkins conditional for `ccache -o max_size=100G` just because I'm not sure **everyone** would want to do that every time the script is run (it creates/modifies a file `$USER/.ccache/ccache.conf` which might not be desirable) and in any case the users of this script are (obviously) capable of running that command themselves. Instead, for non-Jenkins users, just print the current max_size setting and inform them that setting it higher might help improve performance.

Fixes: http://tracker.ceph.com/issues/24777

ErwanAliasr1 commented on run-make-check.sh:
```diff
@@ -76,28 +76,26 @@ function run() {
+    # The following settings are made for improving ccache efficiency
+    # by removing the entropy generated by the date/time embedded in the build
+    CMAKE_BUILD_OPTS="$CMAKE_BUILD_OPTS -D ENABLE_GIT_VERSION=OFF"
+    export SOURCE_DATE_EPOCH=$(date +%D |date -f- +%s)
```
moving SOURCE_DATE_EPOCH and disabling the GIT versioning out of the JENKINS condition could be dangerous if people are not aware of it as the resulting binaries have a broken tracability.

smithfarm replied:
The purpose of `run-make-check.sh` is to make it easy for folks to run "make check" and, in my mind, that includes making it easy for them to run it as fast as possible. If these ccache tweaks bring an order-of-magnitude speed-up, that's something users need to have by default.

ErwanAliasr1 replied:
Why not but in that case we have to warn them explicitly that the generated binaries doesn't have the proper tractability for a production use. Printing a visible message would be fine for me.

smithfarm replied:
Visible message added - please take a look.

ErwanAliasr1 commented on run-make-check.sh:
```diff
@@ -76,28 +76,26 @@ function run() {
+    # by removing the entropy generated by the date/time embedded in the build
+    CMAKE_BUILD_OPTS="$CMAKE_BUILD_OPTS -D ENABLE_GIT_VERSION=OFF"
+    export SOURCE_DATE_EPOCH=$(date +%D |date -f- +%s)
+    ccache -o sloppiness=time_macros
```
Does the user not using ccache for anything else then ceph ? If user had a particular setup, we'll break it. That's why I let that changes in the JENKINS condition.

ErwanAliasr1 replied:
Today not being able to install ccache is not fatal. So it may worth setting a bash -e to insure that we never reach that code out of ccache presence.

smithfarm commented:
@ErwanAliasr1 makes a valid point that the ccache tweaks are persisted via `$USER/.ccache/ccache.conf` which could conceivably break a user's existing ccache setup.

For now, I added some informative messages to the script to bring this matter to the user's attention. Another idea is to implement a `--no-ccache-tweaks` command-line option, but my gut feeling is that 100% of sane users will want to benefit from these ccache tweaks whenever they run the script.

ErwanAliasr1 commented on run-make-check.sh:
```diff
@@ -76,29 +84,34 @@ function run() {
+    echo "Note that this script modifies $USER/.ccache/ccache.conf to ensure maximum performance."
+    echo "If that is not desirable, consider running the script as a different user."
+    echo "Also note that the binaries produced by this script do not contain correct time"
+    echo "and git version information, which may make them unsuitable for debugging."
```
and production use.

smithfarm replied:
Done.

LenzGr commented on run-make-check.sh:
```diff
@@ -76,29 +84,34 @@ function run() {
+    echo "and git version information, which may make them unsuitable for debugging."
+    # remove the entropy generated by the date/time embedded in the build
+    CMAKE_BUILD_OPTS="$CMAKE_BUILD_OPTS -D ENABLE_GIT_VERSION=OFF"
+    export SOURCE_DATE_EPOCH=$(date +%D |date -f- +%s)
```
If I get this correctly, it will maintain better cache utilization for 24 hours, correct? `SOURCE_DATE_EPOCH` only changes daily, which would then result in a rebuild that won't benefit from any previously cached items.

ErwanAliasr1 replied:
True. That was my first expectation for the CI because of the massive rebuilds.
We could extend it yes, if we set it to everyone.
A week ? A month ?

smithfarm replied:
The longer the better. Would there be any major drawback if we set it to `"0"`?

LenzGr replied:
Good question, I think at least a week would be useful. Also please note that @robbat2 [raised a good point](https://www.spinics.net/lists/ceph-devel/msg41868.html):

> This may fail in some locales where `%D` is not trivially parsable.
>
> Can I suggest this instead:
> `export SOURCE_DATE_EPOCH=$(date -d 'today 00:00 UTC' +%s)`

ErwanAliasr1 replied:
ccache rotate cache by its own. Maybe 0 is a serious option to optimize at most the cache for everyone.

smithfarm commented on run-make-check.sh:
```diff
@@ -76,29 +84,34 @@ function run() {
+    # remove the entropy generated by the date/time embedded in the build
+    CMAKE_BUILD_OPTS="$CMAKE_BUILD_OPTS -D ENABLE_GIT_VERSION=OFF"
+    export SOURCE_DATE_EPOCH=$(date +%D |date -f- +%s)
+    ccache -o sloppiness=time_macros
```
That would work only until the ccache maintainers change the format of `ccache -p` . . . I wonder if this additional complication is warranted, given we are already warning users that the script tweaks their local ccache settings.

ErwanAliasr1 replied:
That sounds stable for a long while

smithfarm replied:
@ErwanAliasr1 I added a commit that saves the user's existing local ccache configuration (if any) and restores it when leaving the script. That should address your concern.

smithfarm commented:
I have not tested these tweaks myself. @ErwanAliasr1 Is the Jenkins CI already benefiting from the tweaks? In another PR I saw "make check" complete in 50 minutes just now.

ErwanAliasr1 commented:
@smithfarm This feature still needs some tuning to be efficient at 100% but it starts to pay.

1- it needs the PR to be rebased against my patch to get the feature
2- the cache have to be populated once which takes the 55mn
3- VMs dies after 20mn of idle killing the cache. If we build faster, the VM will be sooner in idle making it die sooner if no other PR incomes quickly. Then we trigger a new VM, without cache. This topic is opened with @djgalloway and we have to fine the right tuning on that. We have also to adjust the number of worker nodes to get the most of the caching. In the CI case, the expiration date of the cache isn't really important as VMs will die at some point.
4- if a PR breaks some important headers, cache will be useless

The good news is that some PRs are passing in 14mn like https://jenkins.ceph.com/job/ceph-pull-requests/828/ which means the patch is now functioning properly on its principles.

ErwanAliasr1 commented:
Please note also this have a very positive impact on arm64 like in 
https://jenkins.ceph.com/job/ceph-pull-requests-arm64/19469/console

Jobs are passing in 30mn instead of 55, that's a great save too.

robbat2 commented on run-make-check.sh:
```diff
@@ -76,29 +84,37 @@ function run() {
+EOM
+    # remove the entropy generated by the date/time embedded in the build
+    CMAKE_BUILD_OPTS="$CMAKE_BUILD_OPTS -D ENABLE_GIT_VERSION=OFF"
+    export SOURCE_DATE_EPOCH=$(date +%D |date -f- +%s)
```
Please stop using %D. It's not as portable as you think. coreutils does hardcode it, but it's been locale dependent before. Additionally the variant I offered is more readible: `$(TZ=UTC date -d 'today 00:00' +%s)`

ErwanAliasr1 replied:
@robbat2 Your point have been taken in account and @LenzGr was supporting that vision too. The pending question is more "what should be the default epoch for best performance". I'm wondering if 0 could be the answer meaning that we don't need the date thing.

LenzGr replied:
Setting `SOURCE_DATE_EPOCH` to zero would be fine with me. In the end, this is for speeding up test builds, not for generating production/release builds.

smithfarm commented:
@ErwanAliasr1 My feeling is that the users of `run-make-check.sh` are a pretty specific bunch. They want to be able to run Ceph's make check, and they want to run it as fast as possible. They are willing to have the script tweak the local user's ccache configuration to achieve this. If any user does not like this (and they will know because the script tells them), they can make a dedicated user for running make check.

That said, I will add some code that checks if there is an existing `$HOME/.ccache/ccache.conf` and, if there is, saves it at the beginning and restores it at the end, including a trap to cover abnormal exit from the script.

ErwanAliasr1 commented on run-make-check.sh:
```diff
@@ -12,9 +12,24 @@
-#
-# Return MAX(1, (number of processors / 2)) by default or NPROC
-#
+trap clean_up_after_myself EXIT
```
Trapping ERR would be nice too

tchaikov replied:
i don't think we should trap ERR, as quite a lot of commands could fail here, for example `test -f $ORIGINAL_CCACHE_CONF` might fail.

ErwanAliasr1 replied:
As long as we do tweak the user env and restore it after a build, we shall be able to restore it also when something fails to get a consistent user env.

smithfarm commented:
This passes my local testing. Any more reviews before I merge?

smithfarm commented:
Wow, this really works! The first complete test run took a very long time, but the second one was one or maybe even two orders of magnitude faster!

smithfarm commented:
ChangeLog:

* rebased
* added commit to fix gtest-parallel issue (unrelated to, but dependent upon, this PR)

tchaikov commented on run-make-check.sh:
```diff
@@ -12,9 +12,27 @@
+
+function clean_up_after_myself() {
+    rm -fr ${CEPH_BUILD_VIRTUALENV:-/tmp}/*virtualenv*
+    export SOURCE_DATE_EPOCH="$SAVED_SOURCE_DATE_EPOCH"
```
nit, i don't think user would want to `source` this script.

smithfarm replied:
Ah, now I get it. The `export` only exports to subshells. Thanks, fixing!

tchaikov commented on run-make-check.sh:
```diff
@@ -60,45 +78,63 @@ function run() {
+EOM
+    save_ccache_conf
+    # remove the entropy generated by the date/time embedded in the build
+    CMAKE_BUILD_OPTS="$CMAKE_BUILD_OPTS -D ENABLE_GIT_VERSION=OFF"
```
nit, could take this opportunity to remove the space after `-D` to be consistent with other part of this script.

smithfarm commented:
Changelog:

* dropped the code that was unnecessarily trying to save/restore `SOURCE_DATE_EPOCH` setting in user's environment
* removed space between `-D` and cmake variable

tchaikov commented:
retest this please.

ErwanAliasr1 commented:
@smithfarm Of course it works ! On the CI that divided the build time by 4 !
Increasing the size is mostly interesting on the CI to handle several "versions" of the builds.
A build is usually generating 4GB of cache. So if we have a PR which is changing some stuff in the headers, that's good to keep in cache the other builds so another PR of the former code can still consume it.

ErwanAliasr1 commented:
retest this please

smithfarm commented:
@tchaikov I think this is ready to merge. Do you agree, or would you rather wait until gtest-parallel is fixed for Python 3?

tchaikov commented on run-make-check.sh:
```diff
@@ -60,45 +76,63 @@ function run() {
+        echo "ERROR: ccache could not be installed"
+        exit 1
+    fi
+    if ! type jq > /dev/null 2>&1 ; then
```
no need to install `jq`. it is installed by `mk-build-deps`.

smithfarm replied:
@tchaikov If the user has a local ccache configuration file, the script saves it before changing the local ccache config, and then restores it upon exit. If we do not trap ERR, the saved file does not get restored.

tchaikov commented:
@smithfarm i don't think it's ready to merge. see the comments above.

tchaikov commented:
aside from https://github.com/ceph/ceph/pull/22867/files#r200936339, lgtm

smithfarm commented:
Changelog:

* rebased to address https://github.com/ceph/ceph/pull/22867/files#r200936339
* run with `set -e` from the very beginning -- see efd2ede commit message (tl;dr we `source install-deps.sh` which does `set -e`, so might as well make it explicit)
* make sure we don't fail the script embarrassingly
* install-deps.sh clobbers our `trap` - reinstate it

smithfarm commented:
@tchaikov Done. Thanks for your valuable input on this PR!

smithfarm commented:
Thanks @ErwanAliasr1
